File: backend/schemas.py
"""Схемы запросов и ответов HTTP API (docs/03-data-model.md, п. 6)."""
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field

Level = Literal["CRITICAL", "HIGH", "NORMAL", "LOW"]
SourceKind = Literal["gmail", "telegram"]
ConnState = Literal["off", "active", "reauth"]
MsgStatus = Literal["PROCESSING", "DONE"]
Theme = Literal["dark", "light"]
Density = Literal["spacious", "compact"]
Period = Literal["all", "today", "week", "month"]
SummaryPeriod = Literal["24h", "week", "month"]

PASSWORD_MIN_LENGTH = 8


class Credentials(BaseModel):
    email: EmailStr
    password: str = Field(min_length=PASSWORD_MIN_LENGTH)


class LoginCredentials(BaseModel):
    """На входе пароль не валидируется по длине: старый короткий пароль
    должен получать «неверный email или пароль», а не 422."""

    email: EmailStr
    password: str


class UserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    email: EmailStr
    criteria: str
    theme: Theme
    density: Density
    created_at: datetime


class MeUpdate(BaseModel):
    criteria: str | None = None
    theme: Theme | None = None
    density: Density | None = None
    current_password: str | None = None
    new_password: str | None = Field(default=None, min_length=PASSWORD_MIN_LENGTH)


class MeUpdateResult(BaseModel):
    user: UserOut
    # Сколько сообщений отправлено на переоценку после смены критериев.
    reanalyze_queued: int = 0


class ConnectionOut(BaseModel):
    kind: SourceKind
    state: ConnState
    account: str
    last_sync_at: datetime | None


class TelegramConnect(BaseModel):
    bot_token: str = Field(min_length=10)


class AuthUrl(BaseModel):
    auth_url: str


class MessageBrief(BaseModel):
    """Строка «Отправитель — Тема» с точкой уровня: карточка источника и сводка."""

    id: int
    sender_name: str
    subject: str
    level: Level


class SourceCard(BaseModel):
    kind: SourceKind
    state: ConnState
    account: str
    last_sync_at: datetime | None
    total: int
    unread: int
    distribution: dict[str, int]
    urgent: MessageBrief | None


class MessageOut(BaseModel):
    id: int
    source: SourceKind
    external_id: str
    sender_name: str
    sender_addr: str
    subject: str
    body: str
    received_at: datetime
    is_read: bool
    status: MsgStatus
    level: Level
    level_override: Level | None
    category: str
    deadline_text: str
    needs_reply: bool
    needs_action: bool
    summary: str
    external_url: str
    analyzed_at: datetime | None
    analysis_failed: bool


class MessageList(BaseModel):
    items: list[MessageOut]
    total: int
    unread: int


class LevelUpdate(BaseModel):
    level: Level


class SummaryOut(BaseModel):
    period: SummaryPeriod
    total: int
    distribution: dict[str, int]
    needs_reply: int
    needs_action: int
    top: list[MessageBrief]

File: backend/internal/schemas/schemas.go
// Package view — перевод моделей в схемы ответа. Один формат сообщения
// на все ручки и на SSE (docs/03-data-model.md, п. 6).
package schemas

import (
	"strings"
	"time"

	"personalinbox/internal/postgres"
)

// Time — время в ответе API: ISO без зоны, как отдавал прежний бэкенд.
// Фронт дописывает «Z» сам (frontend/src/lib/format.ts).
type Time time.Time

const jsonLayout = "2006-01-02T15:04:05"

// MarshalJSON пишет момент в UTC без суффикса зоны.
func (t Time) MarshalJSON() ([]byte, error) {
	return []byte(`"` + time.Time(t).UTC().Format(jsonLayout) + `"`), nil
}

// UnmarshalJSON нужен тестам, которые разбирают собственные ответы.
func (t *Time) UnmarshalJSON(raw []byte) error {
	value := strings.Trim(string(raw), `"`)
	parsed, err := time.Parse(jsonLayout, value)
	if err != nil {
		return err
	}
	*t = Time(parsed)
	return nil
}

// TimePtr — nullable-время: null, если метки нет.
func TimePtr(value *time.Time) *Time {
	if value == nil {
		return nil
	}
	converted := Time(*value)
	return &converted
}

// User — профиль: критерии, тема, плотность.
type User struct {
	ID        int64  `json:"id"`
	Email     string `json:"email"`
	Criteria  string `json:"criteria"`
	Theme     string `json:"theme"`
	Density   string `json:"density"`
	CreatedAt Time   `json:"created_at"`
}

// Connection — строка экрана «Источники».
type Connection struct {
	Kind       string `json:"kind"`
	State      string `json:"state"`
	Account    string `json:"account"`
	LastSyncAt *Time  `json:"last_sync_at"`
}

// MessageBrief — строка «Отправитель — Тема» с точкой уровня:
// карточка источника и сводка.
type MessageBrief struct {
	ID         int64  `json:"id"`
	SenderName string `json:"sender_name"`
	Subject    string `json:"subject"`
	Level      string `json:"level"`
}

// SourceCard — карточка уровня 1.
type SourceCard struct {
	Kind         string         `json:"kind"`
	State        string         `json:"state"`
	Account      string         `json:"account"`
	LastSyncAt   *Time          `json:"last_sync_at"`
	Total        int            `json:"total"`
	Unread       int            `json:"unread"`
	Distribution map[string]int `json:"distribution"`
	Urgent       *MessageBrief  `json:"urgent"`
}

// Message — сообщение целиком: и в списке, и в деталях, и в событии SSE.
type Message struct {
	ID             int64   `json:"id"`
	Source         string  `json:"source"`
	ExternalID     string  `json:"external_id"`
	SenderName     string  `json:"sender_name"`
	SenderAddr     string  `json:"sender_addr"`
	Subject        string  `json:"subject"`
	Body           string  `json:"body"`
	ReceivedAt     Time    `json:"received_at"`
	IsRead         bool    `json:"is_read"`
	Status         string  `json:"status"`
	Level          string  `json:"level"`
	LevelOverride  *string `json:"level_override"`
	Category       string  `json:"category"`
	DeadlineText   string  `json:"deadline_text"`
	NeedsReply     bool    `json:"needs_reply"`
	NeedsAction    bool    `json:"needs_action"`
	Summary        string  `json:"summary"`
	ExternalURL    string  `json:"external_url"`
	AnalyzedAt     *Time   `json:"analyzed_at"`
	AnalysisFailed bool    `json:"analysis_failed"`
}

// MessageList — ответ ленты: счётчики нужны подзаголовку
// «N сообщений · M непрочитанных» и пересчитываются по текущим фильтрам.
type MessageList struct {
	Items  []Message `json:"items"`
	Total  int       `json:"total"`
	Unread int       `json:"unread"`
}

// Summary — сводка за период.
type Summary struct {
	Period       string         `json:"period"`
	Total        int            `json:"total"`
	Distribution map[string]int `json:"distribution"`
	NeedsReply   int            `json:"needs_reply"`
	NeedsAction  int            `json:"needs_action"`
	Top          []MessageBrief `json:"top"`
}

// MeUpdateResult — второе поле говорит фронту, сколько сообщений ушло
// на переоценку после смены критериев.
type MeUpdateResult struct {
	User            User `json:"user"`
	ReanalyzeQueued int  `json:"reanalyze_queued"`
}

// UserOut собирает профиль для ответа.
func UserOut(user *postgres.User) User {
	return User{
		ID:        user.ID,
		Email:     user.Email,
		Criteria:  user.Criteria,
		Theme:     user.Theme,
		Density:   user.Density,
		CreatedAt: Time(user.CreatedAt),
	}
}

// ConnectionOut собирает строку источника.
func ConnectionOut(connection *postgres.Connection) Connection {
	return Connection{
		Kind:       connection.Kind,
		State:      connection.State,
		Account:    connection.Account,
		LastSyncAt: TimePtr(connection.LastSyncAt),
	}
}

// MessageOut собирает сообщение. Уровень наружу отдаётся эффективный:
// ручное исправление важнее оценки модели.
func MessageOut(message *postgres.Message) Message {
	var override *string
	if message.LevelOverride != "" {
		value := message.LevelOverride
		override = &value
	}
	return Message{
		ID:             message.ID,
		Source:         message.Kind,
		ExternalID:     message.ExternalID,
		SenderName:     message.SenderName,
		SenderAddr:     message.SenderAddr,
		Subject:        message.Subject,
		Body:           message.Body,
		ReceivedAt:     Time(message.ReceivedAt),
		IsRead:         message.IsRead,
		Status:         message.Status,
		Level:          message.EffectiveLevel(),
		LevelOverride:  override,
		Category:       message.Category,
		DeadlineText:   message.DeadlineText,
		NeedsReply:     message.NeedsReply,
		NeedsAction:    message.NeedsAction,
		Summary:        message.Summary,
		ExternalURL:    message.ExternalURL,
		AnalyzedAt:     TimePtr(message.AnalyzedAt),
		AnalysisFailed: message.AnalysisFailed,
	}
}

// MessageBriefOut собирает короткую строку сообщения.
func MessageBriefOut(message *postgres.Message) MessageBrief {
	return MessageBrief{
		ID:         message.ID,
		SenderName: message.SenderName,
		Subject:    message.Subject,
		Level:      message.EffectiveLevel(),
	}
}
